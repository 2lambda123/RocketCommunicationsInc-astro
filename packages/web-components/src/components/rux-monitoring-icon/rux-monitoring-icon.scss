:host {
    display: inline-flex;
    padding: 0;
}

:host([hidden]) {
    display: none;
}

*,
*::before,
*::after {
    box-sizing: border-box;
}

.rux-advanced-status {
    display: flex;
    flex-direction: column;
    justify-content: flex-start;
    position: relative;
    margin: var(--spacing-0);
    line-height: 0;
    font-family: var(--font-body-1-font-family);
    font-size: var(--font-body-1-font-size);
    font-weight: var(--font-body-1-font-weight);
    letter-spacing: var(--font-body-1-letter-spacing);
    -webkit-user-select: none;
    -moz-user-select: none;
    -ms-user-select: none;
    user-select: none;
}

.rux-advanced-status__icon-group {
    display: flex;
    position: relative;
    margin: var(--spacing-0) auto var(--spacing-025);
    width: calc(var(--spacing-16) + var(--spacing-4)); //5rem
    padding-top: var(--spacing-1);
}

rux-icon {
    order: 2;
    margin: 0 auto;
}

.rux-advanced-status__status {
    position: absolute;
    top: 0;
    margin: var(--spacing-050) var(--spacing-2);
}

.rux-advanced-status__badge {
    display: block;
    z-index: 2;
    order: 3;
    position: absolute;
    right: 0;
    top: 0;
    box-shadow: inset 0 0 0 1px var(--color-text-secondary);
    border-radius: var(--radius-base);
    padding: var(--spacing-2) var(--spacing-1); //8px 4px
    color: var(--color-palette-neutral-000);
    font-family: var(--font-body-3-font-family);
    font-size: var(--font-body-3-font-size);
    font-weight: var(--font-body-3-font-weight);
    letter-spacing: var(--font-body-3-letter-spacing);
    text-align: center;
    overflow: hidden;
    text-overflow: ellipsis;
    white-space: nowrap;
    background-color: #000;
}

.rux-advanced-status__label {
    color: var(--color-text-primary);
    font-family: var(--font-body-2-font-family);
    font-size: var(--font-body-2-font-size);
    font-weight: var(--font-body-2-font-weight);
    letter-spacing: var(--font-body-2-letter-spacing);
    line-height: var(--font-body-3-line-height);
    text-align: center;
    text-overflow: ellipsis;
    white-space: nowrap;
    overflow: hidden;
    width: 100%;
    max-width: calc(var(--spacing-24) + var(--spacing-2)); //6.5rem
}

.rux-advanced-status__sublabel {
    font-family: var(--font-body-3-font-family);
    font-size: var(--font-body-3-font-size);
    font-weight: var(--font-body-3-font-weight);
    letter-spacing: var(--font-body-3-letter-spacing);
    line-height: var(--font-body-3-line-height);
    color: var(--gsb-color-text);
    display: block;
}

.rux-advanced-status__status rux-status::part(status) {
    line-height: 0;
}

.rux-advanced-status__hidden {
    display: none;
}

:host([status='off']) {
    rux-icon {
        color: var(--status-symbol-color-fill-off-on-dark);
    }
}

:host([status='standby']) {
    rux-icon {
        color: var(--status-symbol-color-fill-standby-on-dark);
    }
}

:host([status='normal']) {
    rux-icon {
        color: var(--status-symbol-color-fill-normal-on-dark);
    }
}

:host([status='caution']) {
    rux-icon {
        color: var(--status-symbol-color-fill-caution-on-dark);
    }
}

:host([status='serious']) {
    rux-icon {
        color: var(--status-symbol-color-fill-serious-on-dark);
    }
}

:host([status='critical']) {
    rux-icon {
        color: var(--status-symbol-color-fill-critical-on-dark);
    }
}
