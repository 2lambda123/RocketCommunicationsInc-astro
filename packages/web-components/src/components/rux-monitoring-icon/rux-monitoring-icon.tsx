import { Component, h, Element, Prop, Watch } from '@stencil/core'
import { Status, StatusTypes } from '../../common/commonTypes.module'
import MonitoringBadge from '../../common/functional-components/MonitoringBadge/MonitoringBadge'
import MonitoringLabel from '../../common/functional-components/MonitoringLabel'

/**
 * @part container - the components container
 * @part monitoring-badge - The component's notification badge
 * @part monitoring-label - The component's label
 * @part monitoring-sublabel - The component's sublabel
 * @part status-icon - the components status symbol
 */
@Component({
    tag: 'rux-monitoring-icon',
    styleUrl: 'rux-monitoring-icon.scss',
    shadow: true,
})
export class RuxMonitoringIcon {
    @Element() el!: HTMLRuxMonitoringIconElement
    /**
     * Styles the icon according to the Astro Status colors.
     * Valid options are the Astro statuses `critical`, `serious`, `caution`, `normal`, `standby`, and `off`.
     */
    @Prop({ reflect: true }) status: Status = 'normal'

    /**
     * Displays a label below the icon
     */
    @Prop() label!: string

    /**
     * Displays a smaller label underneath the icon label
     */
    @Prop() sublabel?: string

    /**
     * Displays an Astro icon matching this string. For a [full list of available icons,
     * see the Icons section in Astro UXDS Guidelines](https://astrouxds.com/ui-components/icons-and-symbols)
     */
    @Prop() icon: string = 'antenna-transmit'

    /**
     * If provided and greater than `0`, displays an outlined number badge at the bottom right of the icon.
     * Numbers above `9999` are abbreviated to `'10K'` or `'100K'` for numbers in the thousands,
     * `'1.5M'` for millions, `'1.5B'` for billions, and uses `'âˆž'` for one trillion or higher.
     */
    @Prop() notifications: number = 0

    componentDidRender() {
        this.handleNotificatonWidth()
    }

    @Watch('status')
    validateStatus(newValue: string) {
        const statusTypes = {
            off: true,
            standby: true,
            normal: true,
            caution: true,
            serious: true,
            critical: true,
        } as StatusTypes
        if (!statusTypes[newValue]) {
            throw new Error('valid status required')
        }
    }

    handleNotificatonWidth() {
        const badge = this.el.shadowRoot!.querySelector(
            '.rux-advanced-status__badge'
        )
        if (badge) {
            /** Size and position of the icon. */
            const iconRect = this.el.getBoundingClientRect()
            /** Size and position of the badge. */
            const badgeRect = badge.getBoundingClientRect()
            /** Offset between the right-edge of the badge and the right-edge of the icon. */
            const offset = badgeRect.right - iconRect.right
            // if the offset is greater than zero, increase the minimum width of the badge
            if (offset > 0) {
                this.el.style.minWidth = iconRect.width + offset + 'px'
            }
        }
    }

    render() {
        return (
            <div
                id="rux-advanced-status__icon"
                class="rux-advanced-status"
                part="container"
            >
                <div class="rux-advanced-status__icon-group">
                    <div class="rux-advanced-status__status">
                        <rux-status
                            status={this.status}
                            part="status-icon"
                        ></rux-status>
                    </div>

                    <rux-icon
                        icon={this.icon}
                        class={`rux-status--${this.status}`}
                        size="2.5rem"
                        exportparts="icon"
                    ></rux-icon>
                    <MonitoringBadge notifications={this.notifications} />
                </div>

                <MonitoringLabel label={this.label} sublabel={this.sublabel} />
            </div>
        )
    }
}
