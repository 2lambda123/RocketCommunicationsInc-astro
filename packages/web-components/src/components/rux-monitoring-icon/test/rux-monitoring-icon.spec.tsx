import { newSpecPage } from '@stencil/core/testing'
import { RuxMonitoringIcon } from '../rux-monitoring-icon'

describe('rux-monitoring-icon', () => {
    it('renders', async () => {
        const page = await newSpecPage({
            components: [RuxMonitoringIcon],
            html: `<rux-monitoring-icon
        icon="altitude"
        label="Altitude for satellite X"
        sublabel="100000m"
        status="standby"
        notifications="10210"
      ></rux-monitoring-icon>`,
        })
        expect(page.root).toMatchSnapshot()
    })

    it('errors with invalid status', async () => {
        const monitorIcon = new RuxMonitoringIcon()

        expect(() => {
            monitorIcon.validateStatus('')
        }).toThrowError('valid status required')
    })
})
