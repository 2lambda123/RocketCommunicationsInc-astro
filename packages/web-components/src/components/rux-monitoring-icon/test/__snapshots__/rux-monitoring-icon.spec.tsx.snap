// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`rux-monitoring-icon renders 1`] = `
<rux-monitoring-icon icon="altitude" label="Altitude for satellite X" notifications="10210" status="standby" sublabel="100000m">
  <mock:shadow-root>
    <div class="rux-advanced-status" id="rux-advanced-status__icon" part="container" title="10210 Altitude for satellite X 100000m">
      <div class="rux-advanced-status__icon-group">
        <div class="rux-advanced-status__status">
          <rux-status status="standby"></rux-status>
        </div>
        <rux-icon class="rux-status--standby" icon="altitude" size="2.5rem"></rux-icon>
        <div class="rux-advanced-status__badge" part="monitoring-badge">
          10K
        </div>
      </div>
      <div class="rux-advanced-status__label">
        <span part="monitoring-label">
          Altitude for satellite X
        </span>
        <span class="rux-advanced-status__sublabel" part="monitoring-sublabel">
          100000m
        </span>
      </div>
    </div>
  </mock:shadow-root>
</rux-monitoring-icon>
`;
